### PR TITLE
Initial CLI and Loadable Module

### LINKED ISSUE #2: Design a module
Put together a Slicer module for one of the pipeline steps

Creating at least one module will set the trend for how future development of this Slicer work will go.

Relies on https://github.com/brain-microstructure-exploration-tools/abcd-microstructure-pipelines/issues/3 having been started

---

Some thoughts on how to approach this:

Make Slicer scripted CLI modules. These allow you to “wrap” a specific processing step that has a well defined input and output.

This works well if you have a standalone python package with entrypoints (as we plan to do with [abcd-microstructure-pipelines](https://github.com/brain-microstructure-exploration-tools/abcd-microstructure-pipelines))

The individual CLI modules for pipeline steps could be in some kind "ABCD Advanced" cateogry or something like that.

Then we can write a Slicer scripted loadable module (or multiple ones) that give a nice UI and everything.

(These suggestions came from internal document [here](https://docs.google.com/document/d/1V9J7Y0SXnwprNEqlOOKvu5oxRrnkjmbidk2QVew6OCg/edit#heading=h.ompigd7h3kba))

### PR DESCRIPTION
**Pull request by allemangD**
- **ENH: ExtensionWizard and preliminary GenMasks CLI module**
- **ENH: Minimal ABCDEnvironment module to install dependencies**

close #2

### CODE REVIEW

**Comment by allemangD** [comment]
ABCDEnvironment module is minimal for now. This functionality should be moved to an orchestractor module once it's avaliable, perhaps with some smarter inspection of the environment to tell if pip_install is even necessary.

I did place the dependency (currently just `abcd-microstructure-pipelines` from git) in a requirements.txt in anticipation of _some_ resolution to the discussions we've been having with the other Slicer maintainers. See https://github.com/Slicer/Slicer/issues/7697 for context on that.

File: ABCDEnvironment/ABCDEnvironment.py
Comment (ebrahimebrahim) [inline ABCDEnvironment/ABCDEnvironment.py]: Can we rename it to AbcdMicroEnvironment or ABCDMicroEnvironment to match the name of the python package the environment setup is meant for

File: ABCDEnvironment/ABCDEnvironment.py
Comment (ebrahimebrahim) [inline ABCDEnvironment/ABCDEnvironment.py]: Please add myself and James :smile:

File: ABCDEnvironment/ABCDEnvironment.py
Comment (ebrahimebrahim) [inline ABCDEnvironment/ABCDEnvironment.py]: Some unused imports here

...or are they? my IDE seems to think they are used but I don't see where

File: ABCDEnvironment/ABCDEnvironment.py
Comment (ebrahimebrahim) [inline ABCDEnvironment/ABCDEnvironment.py]: This should show a busy cursor so it's clear slicer is doing something and not just frozen (even if you don't check the terminal output).

In the past I have used a context manager like this:
```python
class BusyCursor:
    """
    Context manager for showing a busy cursor.  Ensures that cursor reverts to normal in
    case of an exception.
    """

    def __enter__(self):
        qt.QApplication.setOverrideCursor(qt.Qt.BusyCursor)

    def __exit__(self, exception_type, exception_value, traceback):
        qt.QApplication.restoreOverrideCursor()
        return False
```
so then it'd be
```python
with BusyCursor():
    resolve_dependencies()
```

Also it would be good to get some info at the end. For example a message/error box saying whether environment setup succeeded or failed.

(I know we are thinking of this module as a temporary workaround... but realistically we might be stuck with it for a while so it is good to ensure some minimal reasonableness of user experience)

File: ABCDEnvironment/Resources/UI/ABCDEnvironment.ui
Comment (ebrahimebrahim) [inline ABCDEnvironment/Resources/UI/ABCDEnvironment.ui]: Add a label or maybe just parenthesized text to this saying something along the lines of "this may take a few minutes" -- just to make it clear that it's normal for the application to hang

File: ABCDEnvironment/ABCDEnvironment.py
Comment (ebrahimebrahim) [inline ABCDEnvironment/ABCDEnvironment.py]: I like this approach with `importlib.resources`! Never seen this before. I like the way you put a requirements.txt as a resource under `LibABCDEnvironment` :+1: :+1: :+1:

File: GenMasks/GenMasks.py
Comment (ebrahimebrahim) [inline GenMasks/GenMasks.py]: Should be `abcdmicro`, due to https://github.com/brain-microstructure-exploration-tools/abcd-microstructure-pipelines/issues/14

File: GenMasks/GenMasks.xml
Comment (ebrahimebrahim) [inline GenMasks/GenMasks.xml:1]: The setup of this CLI module and everything is great.

When I run it I get

```
ModuleNotFoundError: No module named 'abcd_microstructure_pipelines'
```

I think this is fixed by the rename to `abcdmicro` I mentioned in another comment. But when I tried to just do that I still got an error:
```
Usage: GenMasks.py [OPTIONS]
Try 'GenMasks.py --help' for help.

Error: No such option: --returnparameterfile
```

File: SlicerBrainMicrostructure.png
Comment (ebrahimebrahim) [inline SlicerBrainMicrostructure.png:1]: use this instead: https://drive.google.com/file/d/1OFRpM3bKcMWZPaXF4dpONDtL6I3YoKFN/view?usp=drive_link

It is also a placeholder but better to have a unique looking placeholder.

File: ABCDEnvironment/Resources/Icons/ABCDEnvironment.png
Comment (ebrahimebrahim) [inline ABCDEnvironment/Resources/Icons/ABCDEnvironment.png:1]: use this instead: https://drive.google.com/file/d/1OFRpM3bKcMWZPaXF4dpONDtL6I3YoKFN/view?usp=drive_link

It is also a placeholder but better to have a unique looking placeholder.

**Review by ebrahimebrahim — CHANGES_REQUESTED** [review]

File: ABCDEnvironment/ABCDEnvironment.py
Comment (allemangD) [reply ABCDEnvironment/ABCDEnvironment.py]: Makes sense. I'll do `ABCDMicroEnvironment` for VTK capitalization style.

I'll do it in a separate commit after I address the other requested changes so renamed files don't hide everything.

File: ABCDEnvironment/ABCDEnvironment.py
Comment (allemangD) [reply ABCDEnvironment/ABCDEnvironment.py]: `slicer.util.WaitCursor`

I'll also use `slicer.util.showStatusMessage`.

File: GenMasks/GenMasks.xml
Comment (allemangD) [reply GenMasks/GenMasks.xml:1]: The issue is I did not set `<longflag>` options, so Slicer tries to use `--returnparameterfile` to let the CLI describe its own inputs and outputs. Replace the `<name>` tags with `<longflag>`.

File: ABCDEnvironment/ABCDEnvironment.py
Comment (allemangD) [reply ABCDEnvironment/ABCDEnvironment.py]: I make it a habit because it works nicely with `pip` installations and zipimport / zipapp. Although neither of those apply to Slicer (currently) I think it's a good habit anyway.

**Comment by ebrahimebrahim** [comment]
looks like you might have gotten everything -- ready for another review?

**Comment by allemangD** [comment]
Yep, I forgot there's a button for that.

File: CMakeLists.txt
Comment (ebrahimebrahim) [inline CMakeLists.txt]: Let's set some of these at the outset:

EXTENSION_CATEGORY can be "Diffusion"
EXTENSION_CONTRIBUTORS should be the three of us currently on the project
EXTENSION_DESCRIPTION could be "Tools for analyzing brain microstructure development in longitudinal diffusion MRI data"
EXTENSION_ICONURL can point to the same image I mentioned earlier for module icons

File: ABCDMicroEnvironment/ABCDMicroEnvironment.py
Comment (ebrahimebrahim) [inline ABCDMicroEnvironment/ABCDMicroEnvironment.py]: What if it fails? Could it raise an error dialog or show some kind of error status?

**Review by ebrahimebrahim — CHANGES_REQUESTED** [review]
Tested it and worked!

A couple of small changes and then I'll approve

File: CMakeLists.txt
Comment (allemangD) [reply CMakeLists.txt]: oof, I always forget about these when starting a new project. They should not be default.

I'll set these and also update the contents in the ScriptedLoadableModule class.

Do we have a grant number to credit in acknowledgements?

File: ABCDMicroEnvironment/ABCDMicroEnvironment.py
Comment (allemangD) [reply ABCDMicroEnvironment/ABCDMicroEnvironment.py]: `pip_install` throws if it fails. I'll catch that and `slicer.util.errorDisplay`. I don't think there's a way to disable the CLI module if it fails, but it'll at least show a sensible message and log output.

File: CMakeLists.txt
Comment (ebrahimebrahim) [reply CMakeLists.txt]: Yes, grant is NIH 1R21MH132982-01

**Review by ebrahimebrahim — APPROVED** [review]
Thank you!

**Comment by ebrahimebrahim** [comment]
(added "close #2" to pr description)

**Pull request merged**
